### ISSUE TITLE
Loading 'Defined Strings' (after successful analysis) hangs and slows down UI

### ISSUE THREAD

**Issue by patric-r** [NationalSecurityAgency/ghidra#3044]
**Describe the bug**
Loading 'Defined Strings' (after successful analysis) hangs.
It is a fairly large binary with around ~600000 strings.
Status bar is getting stuck at 37% (in my case after waiting 15 minutes), 100% cpu is consumed on one thread/core.

The sand glass is still animated but the whole UI froze.

Memory exhaustion is not the cause (3 GB from 8GB jvm heap consumed). 

**Expected behavior**
Defined strings view should neither hang nor block ghidra

**Environment (please complete the following information):**
 - OS: [e.g. macOS 10.14.2] Windows 10
 - Java Version: [e.g. 11.0] OpenJDK 11.0.2
 - Ghidra Version: [e.g. 9.1.2] 9.2.3
 - Ghidra Origin: [e.g. official ghidra-sre.org distro, third party distro, locally built]  official

**Additional context**

Stacktraces of the problematic thread.
Interesstingly, it uses reflection.

```

Name: Threaded Table Model Update Manager: Defined String Table
State: RUNNABLE
Total blocked: 318.759  Total waited: 8.831

db.buffers.BufferMgr.getBuffer(BufferMgr.java:957)
   - locked db.buffers.BufferMgr@29a3af25
db.NodeMgr.getLongKeyNode(NodeMgr.java:185)
db.LongKeyInteriorNode.getLeafNode(LongKeyInteriorNode.java:403)
db.Table.hasRecord(Table.java:640)
   - locked db.DBHandle@d393b42
ghidra.program.database.map.AddressKeyRecordIterator.<init>(AddressKeyRecordIterator.java:148)
ghidra.program.database.map.AddressKeyRecordIterator.<init>(AddressKeyRecordIterator.java:63)
ghidra.program.database.code.DataDBAdapterV0.getRecordAfter(DataDBAdapterV0.java:84)
ghidra.program.database.code.CodeManager.getDefinedAddressAfter(CodeManager.java:2798)
ghidra.program.database.code.DataDB.computeLength(DataDB.java:167)
ghidra.program.database.code.DataDB.<init>(DataDB.java:75)
ghidra.program.database.code.DataComponent.<init>(DataComponent.java:80)
ghidra.program.database.code.DataDB.getComponent(DataDB.java:223)
ghidra.program.util.DefinedDataIterator$DataComponentIterator.next(DefinedDataIterator.java:235)
ghidra.program.util.DefinedDataIterator.findNext(DefinedDataIterator.java:157)
ghidra.program.util.DefinedDataIterator.hasNext(DefinedDataIterator.java:131)
ghidra.app.plugin.core.strings.ViewStringsTableModel.doLoad(ViewStringsTableModel.java:117)
docking.widgets.table.threaded.ThreadedTableModel.load(ThreadedTableModel.java:177)
docking.widgets.table.threaded.TableUpdateJob.loadData(TableUpdateJob.java:357)
docking.widgets.table.threaded.TableUpdateJob.processState(TableUpdateJob.java:330)
docking.widgets.table.threaded.TableUpdateJob.run(TableUpdateJob.java:129)
docking.widgets.table.threaded.ThreadedTableModelUpdateMgr$ThreadRunnable.run(ThreadedTableModelUpdateMgr.java:415)
java.base@11.0.2/java.lang.Thread.run(Thread.java:834)




Name: Threaded Table Model Update Manager: Defined String Table
State: RUNNABLE
Total blocked: 318.959  Total waited: 8.897

Stack trace: 
db.LongKeyInteriorNode.<init>(LongKeyInteriorNode.java:53)
db.NodeMgr.getLongKeyNode(NodeMgr.java:197)
db.Table.hasRecord(Table.java:640)
   - locked db.DBHandle@d393b42
ghidra.program.database.map.AddressKeyRecordIterator.<init>(AddressKeyRecordIterator.java:148)
ghidra.program.database.map.AddressKeyRecordIterator.<init>(AddressKeyRecordIterator.java:63)
ghidra.program.database.code.InstDBAdapterV1.getRecordAfter(InstDBAdapterV1.java:115)
ghidra.program.database.code.CodeManager.getDefinedAddressAfter(CodeManager.java:2799)
ghidra.program.database.code.DataDB.computeLength(DataDB.java:167)
ghidra.program.database.code.DataDB.<init>(DataDB.java:75)
ghidra.program.database.code.DataComponent.<init>(DataComponent.java:80)
ghidra.program.database.code.DataDB.getComponent(DataDB.java:223)
ghidra.program.util.DefinedDataIterator$DataComponentIterator.next(DefinedDataIterator.java:235)
ghidra.program.util.DefinedDataIterator.findNext(DefinedDataIterator.java:157)
ghidra.program.util.DefinedDataIterator.hasNext(DefinedDataIterator.java:131)
ghidra.app.plugin.core.strings.ViewStringsTableModel.doLoad(ViewStringsTableModel.java:117)
docking.widgets.table.threaded.ThreadedTableModel.load(ThreadedTableModel.java:177)
docking.widgets.table.threaded.TableUpdateJob.loadData(TableUpdateJob.java:357)
docking.widgets.table.threaded.TableUpdateJob.processState(TableUpdateJob.java:330)
docking.widgets.table.threaded.TableUpdateJob.run(TableUpdateJob.java:129)
docking.widgets.table.threaded.ThreadedTableModelUpdateMgr$ThreadRunnable.run(ThreadedTableModelUpdateMgr.java:415)
java.base@11.0.2/java.lang.Thread.run(Thread.java:834)

Stack trace: 
java.base@11.0.2/jdk.internal.reflect.Reflection.getCallerClass(Native Method)
java.base@11.0.2/java.lang.Class.newInstance(Class.java:576)
db.Schema.createRecord(Schema.java:268)
db.Schema.createRecord(Schema.java:252)
db.FixedRecNode.getRecord(FixedRecNode.java:197)
db.Table$LongKeyRecordIterator.<init>(Table.java:2052)
db.Table.iterator(Table.java:1800)
   - locked db.DBHandle@d393b42
ghidra.program.database.map.AddressKeyRecordIterator.<init>(AddressKeyRecordIterator.java:147)
ghidra.program.database.map.AddressKeyRecordIterator.<init>(AddressKeyRecordIterator.java:63)
ghidra.program.database.code.DataDBAdapterV0.getRecordAfter(DataDBAdapterV0.java:84)
ghidra.program.database.code.CodeManager.getDefinedAddressAfter(CodeManager.java:2798)
ghidra.program.database.code.DataDB.computeLength(DataDB.java:167)
ghidra.program.database.code.DataDB.<init>(DataDB.java:75)
ghidra.program.database.code.DataComponent.<init>(DataComponent.java:80)
ghidra.program.database.code.DataDB.getComponent(DataDB.java:223)
ghidra.program.util.DefinedDataIterator$DataComponentIterator.next(DefinedDataIterator.java:235)
ghidra.program.util.DefinedDataIterator.findNext(DefinedDataIterator.java:157)
ghidra.program.util.DefinedDataIterator.hasNext(DefinedDataIterator.java:131)
ghidra.app.plugin.core.strings.ViewStringsTableModel.doLoad(ViewStringsTableModel.java:117)
docking.widgets.table.threaded.ThreadedTableModel.load(ThreadedTableModel.java:177)
docking.widgets.table.threaded.TableUpdateJob.loadData(TableUpdateJob.java:357)
docking.widgets.table.threaded.TableUpdateJob.processState(TableUpdateJob.java:330)
docking.widgets.table.threaded.TableUpdateJob.run(TableUpdateJob.java:129)
docking.widgets.table.threaded.ThreadedTableModelUpdateMgr$ThreadRunnable.run(ThreadedTableModelUpdateMgr.java:415)
java.base@11.0.2/java.lang.Thread.run(Thread.java:834)


Name: Threaded Table Model Update Manager: Defined String Table
State: RUNNABLE
Total blocked: 362.954  Total waited: 8.897

Stack trace: 
ghidra.util.datastruct.IntListIndexer.clear(IntListIndexer.java:272)
ghidra.util.datastruct.IntKeyIndexer.clear(IntKeyIndexer.java:140)
ghidra.util.datastruct.IntObjectHashtable.removeAll(IntObjectHashtable.java:108)
db.NodeMgr.releaseNodes(NodeMgr.java:126)
db.Table$LongKeyRecordIterator.<init>(Table.java:2057)
db.Table.iterator(Table.java:1800)
   - locked db.DBHandle@d393b42
ghidra.program.database.map.AddressKeyRecordIterator.hasNext(AddressKeyRecordIterator.java:184)
ghidra.program.database.map.AddressKeyRecordIterator.next(AddressKeyRecordIterator.java:224)
ghidra.program.database.code.InstDBAdapterV1.getRecordAfter(InstDBAdapterV1.java:116)
ghidra.program.database.code.CodeManager.getDefinedAddressAfter(CodeManager.java:2799)
ghidra.program.database.code.DataDB.computeLength(DataDB.java:167)
ghidra.program.database.code.DataDB.<init>(DataDB.java:75)
ghidra.program.database.code.DataComponent.<init>(DataComponent.java:80)
ghidra.program.database.code.DataDB.getComponent(DataDB.java:223)
ghidra.program.util.DefinedDataIterator$DataComponentIterator.next(DefinedDataIterator.java:235)
ghidra.program.util.DefinedDataIterator.findNext(DefinedDataIterator.java:157)
ghidra.program.util.DefinedDataIterator.hasNext(DefinedDataIterator.java:131)
ghidra.app.plugin.core.strings.ViewStringsTableModel.doLoad(ViewStringsTableModel.java:117)
docking.widgets.table.threaded.ThreadedTableModel.load(ThreadedTableModel.java:177)
docking.widgets.table.threaded.TableUpdateJob.loadData(TableUpdateJob.java:357)
docking.widgets.table.threaded.TableUpdateJob.processState(TableUpdateJob.java:330)
docking.widgets.table.threaded.TableUpdateJob.run(TableUpdateJob.java:129)
docking.widgets.table.threaded.ThreadedTableModelUpdateMgr$ThreadRunnable.run(ThreadedTableModelUpdateMgr.java:415)
java.base@11.0.2/java.lang.Thread.run(Thread.java:834)

```

**Comment by astrelsky** [comment]
Is this consistently reproducible and is this a program you have the rights to share?

**Comment by dragonmacher** [comment]
Each run here shows the View Strings table as being the culprit.  This is a known issue and has somewhat (mostly?) been mitigated by 316ab7c5ab.

**Comment by patric-r** [comment]
You can five full thread dumps, each taken after approx. 1 second:

[ghidra_issue_3044_thread_dumps.zip](https://github.com/NationalSecurityAgency/ghidra/files/6502409/ghidra_issue_3044_thread_dumps.zip)

Unfortunately, I am not allowed to share the binary here. 

Which newer release should I try?
Is there a release-candidate / beta version out there?

**Comment by patric-r** [comment]
I'd like to add: 
The first 37% (defined strings status bar) took ~30 seconds. After that time, no progress (according the numbers shown) has been made, even not after hours. It really looks like an infinite loop.

**Comment by dragonmacher** [comment]
>  It really looks like an infinite loop.
Interesting.  I assumed the UI was frozen.  Is the user still painting, but with progress that is not moving?

**Comment by patric-r** [comment]
I have to correct myself: The UI is not frozen but damn slow - at least now I am unsure if this was the case when the issue occurred.

Each input takes around 15 seconds. Still 100% cpu on one core/thread.
Progress bar did not move, the numbers are still the same IMHO (I have now made a screenshot and will compare it later on)

**Comment by dragonmacher** [comment]
(I updated my comment above)

The UI may seem largely unresponsive while this table is loading, due to the fact that it keeps grabbing the database lock.   The UI will also sometimes try to acquire the database lock when painting, causing contention.    Normally this is not an issue.  It sounds like you have found a degenerate case.

**Comment by patric-r** [comment]
Something interessting happened: The string table view loading finished!

UI is responding is back to normal, cpu utilization is down to 0%.
I think we can rule out an infinite loop :)

However, this still looks like an issue albeit not a blocker anymore.

I will do a retest and keep you posted.

**Comment by dragonmacher** [comment]
Avoid closing the table.   If you close it, then when you re-open, it will have to load all over again.

**Comment by dev747368** [comment]
If you look at your "Defined Data" table and sort by size, do you have any large items that involve arrays of structs or other nesting arrangement that eventually leads to/contains string elements?

**Comment by patric-r** [comment]
There are many items in the "Defined Data" table but only one large byte array. What I am seeing, this array doesn't containe neither arrays of structs nor other nested elements/strings.

BTW, the issue is reproducible, however my early stated duration is a little bit too pessimistic - it takes approx. 40 minutes but that's still way too long.

**Comment by astrelsky** [comment]
I ended up unintentionally reproducing this earlier at work. Seems to happen on rather large files. I can profile it tomorrow and say what the bottleneck is. Unfortunately I have no way of getting the results off of my workstation.

**Comment by astrelsky** [comment]
@dev747368 @dragonmacher  The culprit appears to be `ghidra.program.database.code.DataDB.computeLength()` with the subsequent call to `ghidra.program.database.code.CodeManager.getDefinedAddressAfter()` taking most of that time. The two biggest hotspots are `db.NodeMgr.releaseNodes()` (29.8%) and `db.buffers.BufferMgr.getBuffer()` (16.2%)

**Comment by homes410** [comment]
I have a patch to workaround this issue.
It resolves #3013 too. experiment with this patch if you have enough time.
[issue-3044.patch.txt](https://github.com/NationalSecurityAgency/ghidra/files/6510551/issue-3044.patch.txt)

**Comment by astrelsky** [comment]
> I have a patch to workaround this issue.
> The patch resolves #3013 too. experiment with this patch if you have enough time.
> [issue-3044.patch-v2.txt](https://github.com/NationalSecurityAgency/ghidra/files/6510584/issue-3044.patch-v2.txt)
> 

I'll give this a shot tomorrow morning.

**Comment by dragonmacher** [comment]
Closed by this 3b1fdc386d26d82593 and other work.    This commit fixed a large memory leak.
